### PR TITLE
SuperClass for tuning strategies and user input for PredicitveTuning

### LINKED ISSUE #452: SuperClass for tuning strategies and user input for PredicitveTuning
**Description**
Creating  a super class for the tuning strategies to bundle functions and increase readability.

Create user input for the variables _relativeOptimum and _maxTuningIterationsWithoutTest in PredictiveTuning.h.

### PR DESCRIPTION
**Pull request by julianpelloth**
# Description

- [x] Create a Super Class for the tuning strategies
- [x] Integrate PredicitveTuning in AutoPas
- [x] Create user input for  _relativeOptimum and _maxTuningIterationsWithoutTest in PredictiveTuning.h.
- [x] Adapt md-flex to new user input

# Other possibility of the super class

If the _currentConfig variable is in the super class instead of the separate tuning strategies, removeN3Option() can be in the super class, but then the super class can only be used for FullSearch and Predicitve Tuning, because right now the other classes use a different data type for _currentConfig.

I would choose the current implementation, but I wanted to mention the other possibility.

## Related Pull Requests

- none

## Resolved Issues

- [x] fixes #452 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] current tests
- [x] run md-flex

### CODE REVIEW

File: src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                                  const std::set<DataLayoutOption> &allowedDataLayoutOptions,
+                                  const std::set<Newton3Option> &allowedNewton3Options);
+
+  /**
+   * Finds the optimum in an unordered map
+   * @param findOptimumSearchSpace
+   * @return optimum
+   */
+  static inline std::__detail::_Node_const_iterator<std::pair<const Configuration, unsigned long>, false, true>
+  getOptimum(const std::unordered_map<Configuration, size_t, ConfigHash> &findOptimumSearchSpace);
```
Comment (FG-TUM) [inline src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h:78]: this is a good candidate where the return type should just be `auto`

File: src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+  /**
+   * Contains every container option.
```
Comment (FG-TUM) [inline src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h:81]:
```suggestion
   * Contains every allowed container option.
```
maybe also rename the member?

File: tests/testAutopas/tests/selectors/tuningStrategy/PredictiveTuningTest.cpp [hunk 3]
```diff
@@ -10,52 +10,46 @@
...
   EXPECT_THAT(predictiveTuning.getAllowedContainerOptions(),
               ::testing::ElementsAre(autopas::ContainerOption::directSum));
 }
 
 TEST_F(PredictiveTuningTest, testSearchSpaceMoreOptions) {
-  unsigned int iteration = 0;
   autopas::PredictiveTuning predictiveTuning(
       {autopas::ContainerOption::linkedCells}, {1.}, {autopas::TraversalOption::c08}, {autopas::DataLayoutOption::soa},
-      {autopas::Newton3Option::enabled, autopas::Newton3Option::disabled}, iteration);
+      {autopas::Newton3Option::enabled, autopas::Newton3Option::disabled}, 1.2, 5);
```
Comment (FG-TUM) [inline tests/testAutopas/tests/selectors/tuningStrategy/PredictiveTuningTest.cpp:32]: make these numbers a variable that you define in the header

File: src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h (hunk 1 above)
Comment (julianpelloth) [reply src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h:78]: If I use auto I get this error message in the building process and I have no clue why:
/usr/bin/ld: CMakeFiles/runTests.dir/tests/selectors/tuningStrategy/FullSearchTest.cpp.o: in function `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)':
FullSearchTest.cpp:(.text+0x3a0): multiple definition of `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)'; CMakeFiles/runTests.dir/tests/selectors/AutoTunerTest.cpp.o:AutoTunerTest.cpp:(.text+0x2240): first defined here
/usr/bin/ld: CMakeFiles/runTests.dir/tests/selectors/tuningStrategy/PredictiveTuningTest.cpp.o: in function `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)':
PredictiveTuningTest.cpp:(.text+0x3a0): multiple definition of `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)'; CMakeFiles/runTests.dir/tests/selectors/AutoTunerTest.cpp.o:AutoTunerTest.cpp:(.text+0x2240): first defined here
/usr/bin/ld: ../../src/autopas/libautopas.a(TuningStrategyFactory.cpp.o): in function `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)':
TuningStrategyFactory.cpp:(.text+0x360): multiple definition of `autopas::TuningStrategySuperClass::getOptimum(std::unordered_map<autopas::Configuration, unsigned long, autopas::ConfigHash, std::equal_to<autopas::Configuration>, std::allocator<std::pair<autopas::Configuration const, unsigned long> > > const&)'; CMakeFiles/runTests.dir/tests/selectors/AutoTunerTest.cpp.o:AutoTunerTest.cpp:(.text+0x2240): first defined here
collect2: error: ld returned 1 exit status
make[2]: *** [tests/testAutopas/CMakeFiles/runTests.dir/build.make:1054: tests/testAutopas/runTests] Error 1
make[1]: *** [CMakeFiles/Makefile2:920: tests/testAutopas/CMakeFiles/runTests.dir/all] Error 2
make: *** [Makefile:112: all] Error 2

File: examples/md-flexible/parsing/MDFlexConfig.h [hunk 4]
```diff
@@ -85,6 +85,10 @@ class MDFlexConfig {
   unsigned int tuningSamples{3};
   static inline const char *tuningMaxEvidenceStr{"tuning-max-evidence"};
   unsigned int tuningMaxEvidence{10};
+  static inline const char *relativeOptimumRangeStr{"relative-optimum-range"};
+  double relativeOptimumRange{1.2};
+  static inline const char *maxTuningPhasesWithoutTestStr{"max-tuning-phases-without-test"};
```
[the remark this answers is not in the document]
Comment (julianpelloth) [reply examples/md-flexible/parsing/MDFlexConfig.h:90]: Yes, it is enough. There is also a little space in between as you can see in the picture.
mdflex_help

File: src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h (hunk 1 above)
Comment (FG-TUM) [reply src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h:78]: Are you sure you changed the signature at bot the declaration and the implementation of the function?
If I change the declaration (around line 77) to:
```
static inline auto getOptimum( .... );
```
and the implementation (around line 122) to:
```
auto TuningStrategySuperClass::getOptimum( ...
```
it compiles on my machine.

File: src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+#include "TuningStrategyInterface.h"
+#include "autopas/containers/CompatibleTraversals.h"
+#include "autopas/utils/ExceptionHandler.h"
+
+namespace autopas {
+
+/**
+ * Super class for tuning strategies
+ */
+class TuningStrategySuperClass : public TuningStrategyInterface {
```
[the remark this answers is not in the document]
Comment (julianpelloth) [reply src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h:18]: My suggestion would be something like searchSpaceBasedTuningStrategy. That is the primary thing that is different to the other classes.

File: src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h (hunk 1 above)
Comment (julianpelloth) [reply src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h:78]: I forgot `inline`. Now it works.

File: src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (FG-TUM) [reply src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h:18]: Technically all Strategies have some kind of search space. The main difference that I see is how they represent it. The classes derived from this one do it through a set.

[the remark this answers is not in the document]
Comment (julianpelloth) [reply src/autopas/selectors/tuningStrategy/TuningStrategySuperClass.h:18]: Then I would choose `SetSearchSpaceBasedTuningStrategy`.

File: tests/testAutopas/tests/selectors/tuningStrategy/PredictiveTuningTest.cpp [hunk 6]
```diff
@@ -365,15 +372,16 @@ TEST_F(PredictiveTuningTest, testInvalidOptimalSearchSpaceOnce) {
  *      - C08 is constant out of the optimum range (15) and invalid in the third iteration.
  *      - Sliced is constant the optimum (10) and invalid in the third iteration.
  *      - C01 is constant out of the optimum range (20).
- * In the third iteration reselectOptimalSearchSpace should be called twice and C01 should be selected after the tuning
- * phase.
+ * In the third iteration reselectOptimalSearchSpace should be called twice and C01 should be selected after the
+ * tuning1.2, 5 phase.
```
Comment (FG-TUM) [inline tests/testAutopas/tests/selectors/tuningStrategy/PredictiveTuningTest.cpp:376]:
```suggestion
 * tuning phase.
```
?

**Comment by FG-TUM** [comment]
update the branch with the latest master when you are ready.

**Pull request merged**